internal/v4/common_test.go

```diff
@@ -291,6 +293,91 @@ func (s *commonSuite) addRequiredCharms(c *gc.C, bundle charm.Bundle) {
 	}
 }
 
+func (s *commonSuite) assertPut(c *gc.C, url string, val interface{}) {
+	s.assertPut0(c, url, val, false)
+}
+
+func (s *commonSuite) assertPutAsAdmin(c *gc.C, url string, val interface{}) {
+	s.assertPut0(c, url, val, true)
+}
+
+func (s *commonSuite) assertPut0(c *gc.C, url string, val interface{}, asAdmin bool) {
+	body, err := json.Marshal(val)
+	c.Assert(err, gc.IsNil)
+	p := httptesting.JSONCallParams{
+		Handler: s.srv,
+		URL:     storeURL(url),
+		Method:  "PUT",
+		Do:      bakeryDo(nil),
+		Header: http.Header{
+			"Content-Type": {"application/json"},
+		},
+		Body: bytes.NewReader(body),
+	}
+	if asAdmin {
+		p.Username = testUsername
+		p.Password = testPassword
+	}
+	httptesting.AssertJSONCall(c, p)
+}
+
+func (s *commonSuite) assertGet(c *gc.C, url string, expectVal interface{}) {
+	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+		Handler:    s.srv,
+		Do:         bakeryDo(nil),
+		URL:        storeURL(url),
+		ExpectBody: expectVal,
+	})
+}
+
+// assertGetUnauthorized asserts that a GET to the given URL results
+// in an ErrUnauthorized response with the given error message.
+func (s *commonSuite) assertGetUnauthorized(c *gc.C, url, expectMessage string) {
+	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+		Handler:      s.srv,
+		Do:           bakeryDo(nil),
+		Method:       "GET",
+		URL:          storeURL(url),
+		ExpectStatus: http.StatusUnauthorized,
+		ExpectBody: params.Error{
+			Code:    params.ErrUnauthorized,
+			Message: expectMessage,
+		},
+	})
+}
+
+// assertGetUnauthorized asserts that a PUT to the given URL with the
+// given body value results in an ErrUnauthorized response with the given
+// error message.
+func (s *commonSuite) assertPutUnauthorized(c *gc.C, url string, val interface{}, expectMessage string) {
+	body, err := json.Marshal(val)
+	c.Assert(err, gc.IsNil)
+	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+		Handler: s.srv,
+		URL:     storeURL(url),
+		Method:  "PUT",
+		Do:      bakeryDo(nil),
+		Header: http.Header{
+			"Content-Type": {"application/json"},
+		},
+		Body:         bytes.NewReader(body),
+		ExpectStatus: http.StatusUnauthorized,
+		ExpectBody: params.Error{
+			Code:    params.ErrUnauthorized,
+			Message: expectMessage,
+		},
+	})
+}
+
+// doAsUser calls the given function, discharging any authorization
+// request as the given user name.
+func (s *commonSuite) doAsUser(user string, f func()) {
+	old := s.discharge
+	s.discharge = dischargeForUser(user)
```

Comment:
If we get a panic, then SetUpTest will still deal with restoring s.discharge.
But maybe worth doing anyway.